### PR TITLE
Refresh the internal LightningOptimizer state for inspection

### LINKED ISSUE #18279: Trainer's optimizer list and the LightningOptimizers state are not kept in sync when updated
### Bug description

We are keeping an inventory of optimizers under the `Trainer.optimizers` list. These are the original optimizers returned by the users in the `configure_optimizers` hook. Internally, we also maintain a list of "LightningOptimizers" that get returned in `LightningModule.optimizers()`. 

Modifying the state on one of these optimizer lists does not update the other.

### What version are you seeing the problem on?

v2.0

### How to reproduce the bug

```python
import torch
from torch.utils.data import DataLoader, Dataset

from lightning.pytorch import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self, restoring=False):
        super().__init__()
        self.restoring = restoring
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def on_train_start(self):
        if self.restoring:
            # self.optimizers().param_groups[0]['lr'] = 100
            self.trainer.optimizers[0].param_groups[0]['lr'] = 100
        print(self.optimizers())
        print(self.trainer.optimizers)
        print(self.trainer.model.optimizers())

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return loss

    def configure_optimizers(self):
        return torch.optim.Adam(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    model = BoringModel()
    trainer1 = Trainer(max_epochs=1)
    trainer1.fit(model, train_dataloaders=train_data)

    # Restore
    model = BoringModel(restoring=True)
    trainer2 = Trainer(max_epochs=2)
    trainer2.fit(model, train_dataloaders=train_data, ckpt_path=trainer1.checkpoint_callback.best_model_path)


if __name__ == "__main__":
    run()
```

### Error messages and logs

There is no error, but the output shows that the states (in this case learning rate) are different:

```
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    capturable: False
    differentiable: False
    eps: 1e-08
    foreach: None
    fused: None
    lr: 0.1
    maximize: False
    weight_decay: 0
)
[Adam (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    capturable: False
    differentiable: False
    eps: 1e-08
    foreach: None
    fused: None
    lr: 100
    maximize: False
    weight_decay: 0
)]
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0): 2.0+
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0): 2.1.0dev
#- Python version (e.g., 3.9): 3.10
#- OS (e.g., Linux): MacOs
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source): pip
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### More info

_No response_

cc @tchaton @awaelchli

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #18279

Our internal wrapper `LightningOptimizer` can't guarantee to keep 100% in sync with internal variables of the optimizer it wraps. This is not an issue for optimization, because the wrapper is only used to redirect the `step()` call, and optimization always uses the real optimizer. This PR ensures the internal state is updated in case the user accesses the LightningOptimizer through `self.optimizers()` in the LightningModule and inspects its state.

cc @borda

### CODE REVIEW

File: src/lightning/pytorch/CHANGELOG.md [hunk 1]
```diff
@@ -203,6 +206,10 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Ensure that the closure running inside the optimizer step has gradients enabled, even if the optimizer step has it disabled ([#18268](https://github.com/Lightning-AI/lightning/pull/18268))
 
 
+- Fixed an issue that could cause the `LightningOptimizer` wrapper returend by `LightningModule.optimizers()` have different internal state than the optimizer it wraps ([#18280](https://github.com/Lightning-AI/lightning/pull/18280))
```
Comment (awaelchli) [inline src/lightning/pytorch/CHANGELOG.md:209]:
```suggestion
- Fixed an issue that could cause the `LightningOptimizer` wrapper returned by `LightningModule.optimizers()` have different internal state than the optimizer it wraps ([#18280](https://github.com/Lightning-AI/lightning/pull/18280))
```

**Review by Borda — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
